### DISCUSSION TITLE
How Zettlr (not only) stores links (backlinks)?

### DISCUSSION THREAD

**Discussion by ghost** [Zettlr/Zettlr#4743] [Q&A] [answered]
Hello!

I want to know how the Zettelkasten apps store back links. I am going to learn new language (rust) and I would like to implement something like program that reads md file and creates efficient* storage of backlinks. How does Zettlr it does? How do other apps do that?

\* - In the sense of speed

**Comment by nathanlesage** [comment, marked as answer]
Zettlr parses each Markdown file according to the Markdown file parser that also highlights the documents, and extracts the backlinks (alongside some other info) from that parse tree and stores it in a metadata format in cache. This takes a lot of time for initial parsing (which is why the first start of the app after an update can take considerable time, as the cache has to be recreated) but makes everything afterwards insanely fast.

Here's how such a descriptor for a file including backlinks can look like:

```json
{
  "root": false,
  "dir": "<redacted>/Notes",
  "path": "<redacted>/Notes/Data is everything!.md",
  "name": "Data is everything!.md",
  "ext": ".md",
  "size": 776,
  "id": "20191008113713",
  "tags": [
    "ai",
    "data",
    "machine-learning",
    "quote"
  ],
  "links": [
    "§§ Artificial Intelligence"
  ],
  "bom": "",
  "type": "file",
  "wordCount": 74,
  "charCount": 481,
  "modtime": 1572356662000,
  "creationtime": 1572356662000,
  "linefeed": "\n",
  "firstHeading": "Data is everything!",
  "frontmatter": null,
  "modified": false
}
```

**Reply by ghost** [reply]
Thank you!
 
1. Just to be sure, `links` array is an array of the backlinks?
2. I expected that the tags would be stored in separate file... Does It have more sense to store them in the descriptor file instead of having one file per tag which just store all the file names/id's that have the tag?
3. I understand the reason to have `modtime` is to catch if file was edited by another program but what reason to store the `creationtime`?
